File: backend/app/models/organization.py
"""องค์กร ผู้ใช้ สิทธิ์ และผู้รับการยกระดับ — 9 ตาราง

ตรงกับ docs/02-data-model.md v2.0 §2.1, §3, §4.1, §4.2, §5.3
"""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import (
    BigInteger,
    Boolean,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.constants import AuthProvider, ContactKey, values
from app.db.base import Base, SoftDeleteMixin, TimestampMixin


class Company(Base, TimestampMixin):
    __tablename__ = "company"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    code: Mapped[str] = mapped_column(String(20), unique=True, nullable=False)
    name_th: Mapped[str] = mapped_column(String(150), nullable=False)
    name_en: Mapped[str | None] = mapped_column(String(150))
    logo_path: Mapped[str | None] = mapped_column(String(255))
    contact_email: Mapped[str | None] = mapped_column(String(150))
    # ห้ามลบบริษัท ใช้ is_active = false เท่านั้น (02-data-model.md §9.3)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    departments: Mapped[list["Department"]] = relationship(back_populates="company")


class Department(Base, TimestampMixin):
    __tablename__ = "department"
    __table_args__ = (UniqueConstraint("company_id", "name"),)

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    company_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("company.id"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(150), nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    company: Mapped[Company] = relationship(back_populates="departments")


class AppUser(Base, TimestampMixin, SoftDeleteMixin):
    """ใช้ชื่อ app_user เพราะ user เป็นคำสงวนของ PostgreSQL"""

    __tablename__ = "app_user"
    __table_args__ = (
        UniqueConstraint("auth_provider", "external_subject"),
        CheckConstraint(
            f"auth_provider IN {tuple(values(AuthProvider))}", name="auth_provider_valid"
        ),
        # ผู้ใช้ local ต้องมีรหัสผ่าน · ผู้ใช้ SSO ไม่ต้องมี (B-01)
        CheckConstraint(
            "auth_provider <> 'local' OR password_hash IS NOT NULL",
            name="local_user_needs_password",
        ),
        Index("ix_app_user_company_active", "company_id", "is_active"),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    company_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("company.id"), nullable=False
    )
    department_id: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("department.id")
    )
    username: Mapped[str] = mapped_column(String(50), unique=True, nullable=False)
    email: Mapped[str | None] = mapped_column(String(150))
    full_name: Mapped[str] = mapped_column(String(150), nullable=False)
    employee_code: Mapped[str | None] = mapped_column(String(50))
    phone: Mapped[str | None] = mapped_column(String(30))
    job_title: Mapped[str | None] = mapped_column(String(100))

    # ── การยืนยันตัวตน ──
    # nullable เพราะผู้ใช้ SSO ในเฟส 2 จะไม่มีรหัสผ่าน — เตรียมไว้ตั้งแต่แรก
    # จะได้ไม่ต้องทำ migration ที่กระทบทุกแถวภายหลัง (B-01)
    password_hash: Mapped[str | None] = mapped_column(String(255))
    auth_provider: Mapped[str] = mapped_column(
        String(20), default=AuthProvider.LOCAL.value, nullable=False
    )
    external_subject: Mapped[str | None] = mapped_column(String(255))
    must_change_password: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    password_changed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    # ── การล็อกบัญชี ──
    # นโยบาย 3.2: "การปลดล็อกต้องยืนยันตัวตนกับ Service Desk"
    # จึงใช้ธง is_locked ไม่ใช่ locked_until — ปลดเองตามเวลาไม่ได้
    failed_login_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    is_locked: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    locked_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    unlocked_by: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("app_user.id"))

    # เพิ่มค่านี้แล้ว token เก่าทั้งหมดของบัญชีเป็นโมฆะทันที
    # ใช้ตอนปิดบัญชีหรือเปลี่ยน role โดยไม่ต้องรอ access token หมดอายุ 30 นาที (B-02)
    token_version: Mapped[int] = mapped_column(Integer, default=0, nullable=False)

    # บัญชีผู้ดูแลต้องแยกจากบัญชีใช้งานประจำวัน + เปลี่ยนรหัสทุก 90 วัน (นโยบาย 3.2)
    is_admin_account: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    last_login_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    # ใช้ส่งการแจ้งเตือนขาออกเท่านั้น ไม่ใช่ช่องทางรับแจ้ง (SLA 3.2)
    line_user_id: Mapped[str | None] = mapped_column(String(100))

    # ระบุ foreign_keys ชัดเจนเพราะ user_role มี FK มาที่ app_user สองเส้น
    # (user_id = เจ้าของสิทธิ์ · granted_by = คนมอบสิทธิ์)
    user_roles: Mapped[list["UserRole"]] = relationship(
        back_populates="user", foreign_keys="UserRole.user_id"
    )


class PasswordHistory(Base):
    """ห้ามใช้รหัสผ่านซ้ำ (นโยบาย 3.2) — เก็บกี่ชุดกำหนดที่ PASSWORD_HISTORY_SIZE"""

    __tablename__ = "password_history"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    user_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("app_user.id", ondelete="CASCADE"), nullable=False, index=True
    )
    password_hash: Mapped[str] = mapped_column(String(255), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )


class Role(Base):
    __tablename__ = "role"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    code: Mapped[str] = mapped_column(String(30), unique=True, nullable=False)
    name_th: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(String(255))
    is_system: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    permissions: Mapped[list["Permission"]] = relationship(
        secondary="role_permission", back_populates="roles"
    )


class Permission(Base):
    __tablename__ = "permission"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    code: Mapped[str] = mapped_column(String(60), unique=True, nullable=False)
    group_name: Mapped[str] = mapped_column(String(40), nullable=False)
    description: Mapped[str | None] = mapped_column(String(255))

    roles: Mapped[list[Role]] = relationship(
        secondary="role_permission", back_populates="permissions"
    )


class RolePermission(Base):
    __tablename__ = "role_permission"

    role_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("role.id", ondelete="CASCADE"), primary_key=True
    )
    permission_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("permission.id", ondelete="CASCADE"), primary_key=True
    )


class UserRole(Base):
    __tablename__ = "user_role"
    __table_args__ = (UniqueConstraint("user_id", "role_id"),)

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    user_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("app_user.id", ondelete="CASCADE"), nullable=False, index=True
    )
    role_id: Mapped[int] = mapped_column(BigInteger, ForeignKey("role.id"), nullable=False)
    granted_by: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("app_user.id"))
    granted_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    # สิทธิ์ชั่วคราวต้องมีกำหนดสิ้นสุด (นโยบาย 3.3 / SOP-03 ข้อ 6)
    # เฟส 1 ใช้รายงานเตือน · เฟส 2 เพิกถอนอัตโนมัติ
    expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), index=True)

    user: Mapped[AppUser] = relationship(back_populates="user_roles", foreign_keys=[user_id])
    role: Mapped[Role] = relationship()
    scopes: Mapped[list["UserRoleScope"]] = relationship(
        back_populates="user_role", cascade="all, delete-orphan"
    )


class UserRoleScope(Base):
    """ขอบเขตบริษัทของ role นั้น

    ถ้าไม่มีแถวเลย -> ขอบเขตเป็นบริษัทต้นสังกัดของผู้ใช้
    super_admin ไม่ต้องมี scope (เห็นทุกบริษัทเสมอ)
    """

    __tablename__ = "user_role_scope"
    __table_args__ = (UniqueConstraint("user_role_id", "company_id"),)

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    user_role_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("user_role.id", ondelete="CASCADE"), nullable=False, index=True
    )
    company_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("company.id"), nullable=False
    )

    user_role: Mapped[UserRole] = relationship(back_populates="scopes")


class EscalationContact(Base, TimestampMixin):
    """ผูกตำแหน่งในองค์กรกับบัญชีผู้ใช้จริง

    Head of IT / CEO / DPO เป็นตำแหน่ง ไม่ใช่ชุดสิทธิ์ — การเพิ่มเป็น role
    จะทำให้ permission matrix บวมโดยไม่ได้ประโยชน์ (05-… §5.1)
    """

    __tablename__ = "escalation_contact"
    __table_args__ = (
        UniqueConstraint("company_id", "contact_key", "user_id"),
        CheckConstraint(
            f"contact_key IN {tuple(values(ContactKey))}", name="contact_key_valid"
        ),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    # null = ระดับกลุ่ม ใช้เป็น fallback เมื่อบริษัทนั้นไม่ได้กำหนดคนของตัวเอง
    company_id: Mapped[int | None] = mapped_column(BigInteger, ForeignKey("company.id"))
    contact_key: Mapped[str] = mapped_column(String(30), nullable=False, index=True)
    user_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("app_user.id"), nullable=False
    )
    is_primary: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

File: backend/src/db/schema/organization.ts
/**
 * องค์กร ผู้ใช้ สิทธิ์ และผู้รับการยกระดับ — 10 ตาราง
 * ตรงกับ docs/02-data-model.md v2.0 §2.1, §3, §4.1, §4.2, §5.3
 */

import { sql } from 'drizzle-orm';
import {
  bigint,
  bigserial,
  boolean,
  check,
  index,
  integer,
  pgTable,
  timestamp,
  unique,
  varchar,
} from 'drizzle-orm/pg-core';

import { AUTH_PROVIDER, CONTACT_KEY } from '../../common/constants';

/** ตัวช่วยสร้าง CHECK จากรายการค่าที่อนุญาต */
export function inList(col: string, values: readonly string[]): ReturnType<typeof sql> {
  return sql.raw(`${col} IN (${values.map((v) => `'${v}'`).join(', ')})`);
}

const timestamps = {
  createdAt: timestamp('created_at', { withTimezone: true }).defaultNow().notNull(),
  updatedAt: timestamp('updated_at', { withTimezone: true }).defaultNow().notNull(),
};

export const company = pgTable('company', {
  id: bigserial('id', { mode: 'number' }).primaryKey(),
  code: varchar('code', { length: 20 }).notNull().unique(),
  nameTh: varchar('name_th', { length: 150 }).notNull(),
  nameEn: varchar('name_en', { length: 150 }),
  logoPath: varchar('logo_path', { length: 255 }),
  contactEmail: varchar('contact_email', { length: 150 }),
  // ห้ามลบบริษัท ใช้ is_active = false เท่านั้น
  isActive: boolean('is_active').default(true).notNull(),
  ...timestamps,
});

export const department = pgTable(
  'department',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    companyId: bigint('company_id', { mode: 'number' })
      .notNull()
      .references(() => company.id),
    name: varchar('name', { length: 150 }).notNull(),
    isActive: boolean('is_active').default(true).notNull(),
    ...timestamps,
  },
  (t) => [unique('uq_department_company_name').on(t.companyId, t.name), index('ix_department_company').on(t.companyId)],
);

export const appUser = pgTable(
  'app_user',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    companyId: bigint('company_id', { mode: 'number' })
      .notNull()
      .references(() => company.id),
    departmentId: bigint('department_id', { mode: 'number' }).references(() => department.id),
    username: varchar('username', { length: 50 }).notNull().unique(),
    email: varchar('email', { length: 150 }),
    fullName: varchar('full_name', { length: 150 }).notNull(),
    employeeCode: varchar('employee_code', { length: 50 }),
    phone: varchar('phone', { length: 30 }),
    jobTitle: varchar('job_title', { length: 100 }),

    // nullable เพราะผู้ใช้ SSO ในเฟส 2 จะไม่มีรหัสผ่าน — เตรียมไว้ตั้งแต่แรก
    // จะได้ไม่ต้องทำ migration ที่กระทบทุกแถวภายหลัง (B-01)
    passwordHash: varchar('password_hash', { length: 255 }),
    authProvider: varchar('auth_provider', { length: 20 }).default('local').notNull(),
    externalSubject: varchar('external_subject', { length: 255 }),
    mustChangePassword: boolean('must_change_password').default(true).notNull(),
    passwordChangedAt: timestamp('password_changed_at', { withTimezone: true }),

    // นโยบาย 3.2: "การปลดล็อกต้องยืนยันตัวตนกับ Service Desk"
    // จึงใช้ธง is_locked ไม่ใช่ locked_until — ปลดเองตามเวลาไม่ได้
    failedLoginCount: integer('failed_login_count').default(0).notNull(),
    isLocked: boolean('is_locked').default(false).notNull(),
    lockedAt: timestamp('locked_at', { withTimezone: true }),
    unlockedBy: bigint('unlocked_by', { mode: 'number' }),

    // เพิ่มค่านี้แล้ว token เก่าทั้งหมดของบัญชีเป็นโมฆะทันที (B-02)
    tokenVersion: integer('token_version').default(0).notNull(),
    isAdminAccount: boolean('is_admin_account').default(false).notNull(),

    isActive: boolean('is_active').default(true).notNull(),
    lastLoginAt: timestamp('last_login_at', { withTimezone: true }),
    // ใช้ส่งการแจ้งเตือนขาออกเท่านั้น ไม่ใช่ช่องทางรับแจ้ง (SLA 3.2)
    lineUserId: varchar('line_user_id', { length: 100 }),

    deletedAt: timestamp('deleted_at', { withTimezone: true }),
    ...timestamps,
  },
  (t) => [
    unique('uq_app_user_provider_subject').on(t.authProvider, t.externalSubject),
    check('ck_app_user_auth_provider_valid', inList('auth_provider', AUTH_PROVIDER)),
    // ผู้ใช้ local ต้องมีรหัสผ่าน · ผู้ใช้ SSO ไม่ต้องมี
    check(
      'ck_app_user_local_needs_password',
      sql`auth_provider <> 'local' OR password_hash IS NOT NULL`,
    ),
    index('ix_app_user_company_active').on(t.companyId, t.isActive),
  ],
);

/** ห้ามใช้รหัสผ่านซ้ำ (นโยบาย 3.2) — เก็บกี่ชุดกำหนดที่ PASSWORD_HISTORY_SIZE */
export const passwordHistory = pgTable(
  'password_history',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    userId: bigint('user_id', { mode: 'number' })
      .notNull()
      .references(() => appUser.id, { onDelete: 'cascade' }),
    passwordHash: varchar('password_hash', { length: 255 }).notNull(),
    createdAt: timestamp('created_at', { withTimezone: true }).defaultNow().notNull(),
  },
  (t) => [index('ix_password_history_user').on(t.userId)],
);

export const role = pgTable('role', {
  id: bigserial('id', { mode: 'number' }).primaryKey(),
  code: varchar('code', { length: 30 }).notNull().unique(),
  nameTh: varchar('name_th', { length: 100 }).notNull(),
  description: varchar('description', { length: 255 }),
  isSystem: boolean('is_system').default(true).notNull(),
});

export const permission = pgTable('permission', {
  id: bigserial('id', { mode: 'number' }).primaryKey(),
  code: varchar('code', { length: 60 }).notNull().unique(),
  groupName: varchar('group_name', { length: 40 }).notNull(),
  description: varchar('description', { length: 255 }),
});

export const rolePermission = pgTable(
  'role_permission',
  {
    roleId: bigint('role_id', { mode: 'number' })
      .notNull()
      .references(() => role.id, { onDelete: 'cascade' }),
    permissionId: bigint('permission_id', { mode: 'number' })
      .notNull()
      .references(() => permission.id, { onDelete: 'cascade' }),
  },
  (t) => [unique('pk_role_permission').on(t.roleId, t.permissionId)],
);

export const userRole = pgTable(
  'user_role',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    userId: bigint('user_id', { mode: 'number' })
      .notNull()
      .references(() => appUser.id, { onDelete: 'cascade' }),
    roleId: bigint('role_id', { mode: 'number' })
      .notNull()
      .references(() => role.id),
    grantedBy: bigint('granted_by', { mode: 'number' }),
    grantedAt: timestamp('granted_at', { withTimezone: true }).defaultNow().notNull(),
    // สิทธิ์ชั่วคราวต้องมีกำหนดสิ้นสุด (นโยบาย 3.3 / SOP-03 ข้อ 6)
    expiresAt: timestamp('expires_at', { withTimezone: true }),
  },
  (t) => [
    unique('uq_user_role').on(t.userId, t.roleId),
    index('ix_user_role_user').on(t.userId),
    index('ix_user_role_expires').on(t.expiresAt),
  ],
);

/**
 * ขอบเขตบริษัทของ role นั้น
 * ถ้าไม่มีแถวเลย → ขอบเขตเป็นบริษัทต้นสังกัดของผู้ใช้
 * super_admin ไม่ต้องมี scope (เห็นทุกบริษัทเสมอ)
 */
export const userRoleScope = pgTable(
  'user_role_scope',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    userRoleId: bigint('user_role_id', { mode: 'number' })
      .notNull()
      .references(() => userRole.id, { onDelete: 'cascade' }),
    companyId: bigint('company_id', { mode: 'number' })
      .notNull()
      .references(() => company.id),
  },
  (t) => [
    unique('uq_user_role_scope').on(t.userRoleId, t.companyId),
    index('ix_user_role_scope_role').on(t.userRoleId),
  ],
);

/**
 * ผูกตำแหน่งในองค์กรกับบัญชีผู้ใช้จริง
 *
 * Head of IT / CEO / DPO เป็นตำแหน่ง ไม่ใช่ชุดสิทธิ์ — การเพิ่มเป็น role
 * จะทำให้ permission matrix บวมโดยไม่ได้ประโยชน์ (05-… §5.1)
 */
export const escalationContact = pgTable(
  'escalation_contact',
  {
    id: bigserial('id', { mode: 'number' }).primaryKey(),
    // null = ระดับกลุ่ม ใช้เป็น fallback เมื่อบริษัทนั้นไม่ได้กำหนดคนของตัวเอง
    companyId: bigint('company_id', { mode: 'number' }).references(() => company.id),
    contactKey: varchar('contact_key', { length: 30 }).notNull(),
    userId: bigint('user_id', { mode: 'number' })
      .notNull()
      .references(() => appUser.id),
    isPrimary: boolean('is_primary').default(true).notNull(),
    isActive: boolean('is_active').default(true).notNull(),
    ...timestamps,
  },
  (t) => [
    unique('uq_escalation_contact').on(t.companyId, t.contactKey, t.userId),
    check('ck_escalation_contact_key_valid', inList('contact_key', CONTACT_KEY)),
    index('ix_escalation_contact_key').on(t.contactKey),
  ],
);
